test/UnitTests/MSTest.Analyzers.UnitTests/DynamicDataShouldBeValidAnalyzerTests.cs

```diff
@@ -1356,4 +1356,30 @@ await VerifyCS.VerifyAnalyzerAsync(
             VerifyCS.Diagnostic(DynamicDataShouldBeValidAnalyzer.DisplayMethodSignatureRule).WithLocation(10).WithArguments("SomeClass", "GetSomeDisplayName3"),
             VerifyCS.Diagnostic(DynamicDataShouldBeValidAnalyzer.DisplayMethodSignatureRule).WithLocation(11).WithArguments("SomeClass", "GetSomeDisplayName3"));
     }
+
+    [TestMethod]
+    [GitHubWorkItem("https://github.com/microsoft/testfx/issues/4922")]
+    public async Task WhenDataImplementsIEnumerable_NoDiagnostic()
+        => await VerifyCS.VerifyAnalyzerAsync("""
+            using System.Collections.Generic;
+            using Microsoft.VisualStudio.TestTools.UnitTesting;
+            [TestClass]
+            public class MyTestClass
+            {
+                [TestMethod]
+                [DynamicData(nameof(GetData), DynamicDataSourceType.Method)]
+                public void TestMethod2(int i, int j)
+                {
+                }
+
+                private static List<object[]> GetData()
+                {
+                    return new List<object[]>
+                    {
+                        new object[] { 1, 2 },
+                        new object[] { 3, 4 },
+                    };
+                }
```

Comment:
Hypothetical scenario, what if the return type is `object` while the actual implementation returns `List`.

I think this is allowed but will be a false positive (and nothing much we can do about it).

Probably not for this PR, but maybe in future we can use two diagnostic IDs? One when we are confident that this will break something, and another when we are less confident which is info by default.

Or maybe in future we can add some dataflow analysis 👀

Reply:
Added a test to show we report. We can improve later if we want.

I do like the idea of multiple IDs based on confidence but I'll consider this out of the scope of this PR.